### PR TITLE
feat: pre-push hook and CI gas artifacts

### LINKED ISSUE #144: feat: create pre-push hook updating gas snapshot file
# Feature Request

**Describe the Feature Request**
create a pre-push hook that runs the `scripts/gen_snapshot.py` script to update the `gas_snapshot.json` file with the current values
**Describe Preferred Solution**

**Related Code**

**Additional Context**

**If the feature request is approved, would you be willing to submit a PR?**
_(Help can be provided if you need assistance submitting a PR)_

- [x] Yes
- [ ] No

### LINKED ISSUE #145: feat: upload gas snapshot to gh artifacts rather than committing them
# Feature Request

**Describe the Feature Request**

Feat: upload gas snapshot to gh artifacts rather than committing them

**Describe Preferred Solution**

1. Upload gas snapshots upon merging on main in the CI.
2. When running compare gas consumption script, download from the gh artifacts the latest gas snapshot.

**Related Code**

**Additional Context**

Currently, we commit gas snapshots, this'll need to conflicts in the future, as we've seen on kakarot cairo0

**If the feature request is approved, would you be willing to submit a PR?**
_(Help can be provided if you need assistance submitting a PR)_

- [x] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by enitrat**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: Closes #144 Closes #145 

## What is the new behavior?

- Updated the compare_snapshot script to download the previous snapshot from artifacts
- Created a pre-push hook that runs the snapshot comparison
- CI now generates gas snapshot artifacts on pushes to the main branch

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Added changes to [`CHANGELOG.md`](../CHANGELOG.md)

- [x] Yes
- [ ] No

### CODE REVIEW

**Review by Eikix — COMMENTED** [review]
nice PR! we should probably add gas_snapshot.json in gitignore

File: .github/workflows/gas_snapshot.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+name: Generate and Upload Gas Snapshot
+
+on:
```
Comment (Eikix) [inline .github/workflows/gas_snapshot.yml:3]: did you try it by pushing your local branch to your fork's main

File: scripts/install_hook.sh [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
```
Comment (Eikix) [inline scripts/install_hook.sh:1]: cool hook

Comment (Eikix) [reply scripts/install_hook.sh:1]: did you try it as well?

File: .github/workflows/gas_snapshot.yml (hunk 1 above)
Comment (enitrat) [reply .github/workflows/gas_snapshot.yml:3]: no, good idea, will do

File: scripts/install_hook.sh (hunk 2 above)
Comment (enitrat) [reply scripts/install_hook.sh:1]: yes

File: scripts/compare_snapshot.py [hunk 3]
```diff
@@ -1,13 +1,82 @@
...
+                if key == "GITHUB_TOKEN":
+                    return value
+    except FileNotFoundError:
+        print(f"Error: {file_path} file not found.")
+    except ValueError:
+        print(f"Error: Invalid format in {file_path}. Expected 'KEY=VALUE' format.")
+    return None
+
+def get_previous_snapshot():
+    REPO = "enitrat/kakarot-ssj"  # Replace with your GitHub username and repo name
```
Comment (Eikix) [inline scripts/compare_snapshot.py:28]: this can't be kkrt-labs/ ? otherwise this also creates git conflicts

**Review by Eikix — COMMENTED** [review]
overall lgtm

File: scripts/compare_snapshot.py (hunk 3 above)
Comment (enitrat) [reply scripts/compare_snapshot.py:28]:
```suggestion
    REPO = "kkrt-labs/kakarot-ssj"  # Replace with your GitHub username and repo name
```

**Review by Eikix — COMMENTED** [review]
lgtm!

**Pull request merged**